src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -693,6 +694,12 @@ object PruneDeadFields {
         val expr2 = rebuild(expr, child2.typ, memo)
         val newKey2 = rebuild(newKey, child2.typ, memo)
         TableKeyByAndAggregate(child2, expr2, newKey2, nPartitions, bufferSize)
+      case LocalizeEntries(child, fieldName) =>
+        val child2 = rebuild(child, memo)
+        if (child2.typ.rvRowType.fields.contains(MatrixType.entriesIdentifier))
+          LocalizeEntries(child2, fieldName)
+        else
+          MatrixRowsTable(child2)
```

Comment:
It appears that this rewrite pass causes assertion failures as we assert that optimization doesn't change types in [Optimize](https://github.com/chrisvittal/hail/blob/373a83d486401a920764b1d3e109fe94a042f2fd/src/main/scala/is/hail/expr/ir/Optimize.scala#L14). Thoughts?

Reply:
What triggers the failure?

Reply:
also, this if will _always_ take the first branch -- rvRowType always has an entry array, which might be `array<struct{}>`

Reply:
The first hit was on the assertion in optimize that I linked above. The issue is very much the rebuild rule for LocalizeEntries. If it is commented out, the tests pass again.

Reply:
The assertion in Optimize is only for the top-level IR -- nested children can change type as much as necessary.

We need to add tests to PruneSuite for the memoization and the rebuilding (should test that something without dependence on  `fieldName` emits a MatrixRowsTable)

Reply:
which test caused your error?

Reply:
`[doctest] hail.matrixtable.MatrixTable.index_entries`

Full message
```
_____________ [doctest] hail.matrixtable.MatrixTable.index_entries _____________
[gw0] linux -- Python 3.6.5 /home/BROAD.MIT.EDU/cvittal/.local/opt/anaconda/envs/hail/bin/python
2352 Expose the entries as if looked up in a dictionary, indexing
2353         with `exprs`.
2354 
2355         Examples
2356         --------
2357         >>> dataset_result = dataset.annotate_entries(GQ2 = dataset2.index_entries(dataset.row_key, dataset.col_key).GQ)
UNEXPECTED EXCEPTION: FatalError('AssertionError: assertion failed: optimization changed type!\n  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}\n  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}\n\nJava stack trace:\njava.lang.AssertionError: assertion failed: optimization changed type!\n  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}\n  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}\n\tat scala.Predef$.assert(Predef.scala:170)\n\tat is.hail.expr.ir.Optimize$.optimize(Optimize.scala:14)\n\tat is.hail.expr.ir.Optimize$.apply(Optimize.scala:21)\n\tat is.hail.table.Table.value$lzycompute(Table.scala:214)\n\tat is.hail.table.Table.value(Table.scala:213)\n\tat is.hail.table.Table.x$5$lzycompute(Table.scala:218)\n\tat is.hail.table.Table.x$5(Table.scala:218)\n\tat is.hail.table.Table.rvd$lzycompute(Table.scala:218)\n\tat is.hail.table.Table.rvd(Table.scala:218)\n\tat is.hail.variant.MatrixTable.annotateRowsTable(MatrixTable.scala:737)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:280)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:214)\n\tat java.lang.Thread.run(Thread.java:748)\n\nHail version: devel-846e87983b0a\nError summary: AssertionError: assertion failed: optimization changed type!\n  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}\n  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}',)
Traceback (most recent call last):

  File "/home/BROAD.MIT.EDU/cvittal/.local/opt/anaconda/envs/hail/lib/python3.6/doctest.py", line 1330, in __run
    compileflags, 1), test.globs)

  File "<doctest hail.matrixtable.MatrixTable.index_entries[0]>", line 1, in <module>

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 975, in annotate_entries
    return self._select_entries(caller, s=self.entry.annotate(**e))

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/typecheck/check.py", line 547, in wrapper
    return f(*args_, **kwargs_)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 2886, in _select_entries
    base, cleanup = self._process_joins(s)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 2545, in _process_joins
    return process_joins(self, exprs, broadcast_f)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/utils/misc.py", line 354, in process_joins
    left = j.join_func(left)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/typecheck/check.py", line 305, in f
    ret = x(*args)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 2456, in joiner
    col_exprs = {col_uid: src_cols_indexed.index(*col_exprs)[col_uid]})

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/typecheck/check.py", line 547, in wrapper
    return f(*args_, **kwargs_)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 2484, in _annotate_all
    base, cleanup = self._process_joins(*all_exprs)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py", line 2545, in _process_joins
    return process_joins(self, exprs, broadcast_f)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/utils/misc.py", line 354, in process_joins
    left = j.join_func(left)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/typecheck/check.py", line 305, in f
    ret = x(*args)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/table.py", line 1396, in joiner
    return MatrixTable(left._jvds.annotateRowsTable(right._jt, uid, False))

  File "/home/BROAD.MIT.EDU/cvittal/.local/opt/spark/spark-2.2.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
    answer, self.gateway_client, self.target_id, self.name)

  File "/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/utils/java.py", line 200, in deco
    'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None

hail.utils.java.FatalError: AssertionError: assertion failed: optimization changed type!
  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}
  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}

Java stack trace:
java.lang.AssertionError: assertion failed: optimization changed type!
  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}
  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}
        at scala.Predef$.assert(Predef.scala:170)
        at is.hail.expr.ir.Optimize$.optimize(Optimize.scala:14)
        at is.hail.expr.ir.Optimize$.apply(Optimize.scala:21)
        at is.hail.table.Table.value$lzycompute(Table.scala:214)
        at is.hail.table.Table.value(Table.scala:213)
        at is.hail.table.Table.x$5$lzycompute(Table.scala:218)
        at is.hail.table.Table.x$5(Table.scala:218)
        at is.hail.table.Table.rvd$lzycompute(Table.scala:218)
        at is.hail.table.Table.rvd(Table.scala:218)
        at is.hail.variant.MatrixTable.annotateRowsTable(MatrixTable.scala:737)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:280)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.GatewayConnection.run(GatewayConnection.java:214)
        at java.lang.Thread.run(Thread.java:748)

Hail version: devel-846e87983b0a
Error summary: AssertionError: assertion failed: optimization changed type!
  before: Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64},__uid_60:Array[Struct{GT:Call,AD:Array[+Int32],DP:Int32,GQ:Int32,PL:Array[+Int32]}]}}
  after:  Table{global:Struct{global_field_1:Int32,global_field_2:Int32,pli:Dict[String,Float64],populations:Array[String],global_field:Int32},key:[locus,alleles],row:Struct{locus:Locus(GRCh37),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{NEGATIVE_TRAIN_SITE:Boolean,HWP:Float64,AC:Array[Int32],culprit:String,MQ0:Int32,ReadPosRankSum:Float64,AN:Int32,InbreedingCoeff:Float64,AF:Array[Float64],GQ_STDDEV:Float64,FS:Float64,DP:Int32,GQ_MEAN:Float64,POSITIVE_TRAIN_SITE:Boolean,VQSLOD:Float64,ClippingRankSum:Float64,BaseQRankSum:Float64,MLEAF:Array[Float64],MLEAC:Array[Int32],MQ:Float64,QD:Float64,END:Int32,DB:Boolean,HaplotypeScore:Float64,MQRankSum:Float64,CCC:Int32,NCC:Int32,DS:Boolean},use_in_kinship:Boolean,panel_maf:Float64,anno1:Int32,anno2:Int32,consequence:String,gene:Array[String],score:Float64,a_index:Int32,variant_qc:Struct{AC:Array[Int32],AF:Array[Float64],AN:Int32,homozygote_count:Array[Int32],dp_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},gq_stats:Struct{mean:Float64,stdev:Float64,min:Float64,max:Float64},n_called:Int64,n_not_called:Int64,call_rate:Float32,n_het:Int64,n_non_ref:Int64,r_expected_het_freq:Float64,p_hwe:Float64}}}

/home/BROAD.MIT.EDU/cvittal/src/hail2/build/tmp/doctest/python/hail/matrixtable.py:2357: UnexpectedException
============== 1 failed, 447 passed, 14 skipped in 86.27 seconds ===============
:doctest FAILED
```